File: src/features/typing/data/go/validate_binary_search_tree.go
package main

import "math"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isValidBST(root *TreeNode) bool {
	var validate func(node *TreeNode, min, max int) bool
	validate = func(node *TreeNode, min, max int) bool {
		if node == nil {
			return true
		}
		if node.Val <= min || node.Val >= max {
			return false
		}
		return validate(node.Left, min, node.Val) && validate(node.Right, node.Val, max)
	}
	return validate(root, math.MinInt, math.MaxInt)
}

File: src/features/typing/data/dart/validate_binary_search_tree.dart
class Solution {
  bool isValidBST(TreeNode? root) {
    bool validate(TreeNode? node, int? min, int? max) {
      if (node == null) {
        return true;
      }
      if (min != null && node.val <= min) {
        return false;
      }
      if (max != null && node.val >= max) {
        return false;
      }
      return validate(node.left, min, node.val) && validate(node.right, node.val, max);
    }
    return validate(root, null, null);
  }
}
